### DISCUSSION TITLE
Debug Renderer example

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#49] [Q&A] [answered]
I am integrating Jolt to PlayCanvas and one thing I am missing dearly is a debug renderer. I can see it in the sources, but looks like it is not exposed in IDL. It would be great to have an IDL variant that includes it and an example of how it can be used.

In case of Ammo, we make different IDL variants that allow us to make different builds. E.g., production, dev, light (excludes some exotic features, like soft bodies, character controller, debug renderer, etc to save on size). The reason for shaving off the features is that web apps often run in size sensitive environments, e.g. Facebook Instant Games, where you are limited to total app size. Unlike native, the web apps are also critical to the loading times, which heavily depend the quality of the network connection of the target device. A default IDL (the one in the repo) could contain all features, similar like Ammo does it, and the user can customize it (remove features he doesn't need), when needed.

**Comment by LeXXik** [comment]
Here is an example of Ammo debug renderer:
Project: https://playcanvas.com/project/744419/overview/ammo-debug-draw
Play: https://playcanv.as/p/OEVC2poo/

Basically, Ammo just calls a user provided Javascript closure and passes it points for a line to be drawn and its color. Then its up to the framework to use that data to draw lines.

In contrast, Rapier physics engine just gives 2 buffers on request (vertex buffer and color buffer). Then a framework uses those to draw lines. I find it more performant, as calling a JS closure for each line from Wasm is a bit expensive.

**Comment by jrouwe** [comment]
It would indeed be possible to expose the ```DebugRenderer``` (example of what you have to implement [here](https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Renderer/DebugRendererRecorder.cpp)). Most of the stuff that is drawn goes through ```DebugRenderer::DrawGeometry``` with only a few calls to ```DebugRenderer::DrawLine``` and ```DebugRenderer::DrawTriangle``` so I think it would be fast enough. In the current build however ```JPH_DEBUG_RENDERER``` is not defined so the class doesn't exist. This would indeed mean that several builds will need to be created and I'm not really sure how that would work with the [WebIDL](https://github.com/jrouwe/JoltPhysics.js/blob/main/JoltJS.idl) interface I'm using (I don't think you can leave out interfaces in certain builds).

Can I ask what you would want from this interface? Because if you just want to draw the bodies, then you can use the [existing](https://github.com/jrouwe/JoltPhysics.js/blob/f3b759a880445a23f4d415db6835992d28579978/Examples/js/example.js#L191) functionality to extract triangles from a shape to render them with very little JS code.

**Comment by LeXXik** [comment]
Well, the reason is very simple - I want to see the current state of the physics world. The way the physics engine sees it, rather than the rendering framework's representation via its meshes. Pretty often we have some collision issues, where the visual mesh does not match the physical body transforms. This renderer helps tremendously debugging such cases.

To be honest I am a complete newb when it comes to Jolt :D So, I wasn't aware of that triangles method. This looks very promising and if I understand it correctly, then it should solve the need just fine. I will give it a go.

**Comment by LeXXik** [comment]
Tackling the debug renderer now. I prefer a single source of truth, so instead of iterating a user map of bodies, I'd like to ask Jolt to give me the list to loop through. Is there a way to make a query to get all bodies (only active? only static/dynamic?)? Even if it is a slow method, as it would be used for debugging anyway.

**Comment by jrouwe** [comment]
There is:

https://github.com/jrouwe/JoltPhysics/blob/02939077a1d13d9704b550d847955a53f6b4f235/Jolt/Physics/PhysicsSystem.h#L173

but it's not exposed in the idl file yet. I can add this (but not today).

**Reply by LeXXik** [reply]
Ooh! That one and the one for active bodies would be great to have! I am computing a motion state for active dynamic bodies between simulation steps. The `GetActiveBodies` would come handy for that 👍 Oh, and no worries, whenever you are convenient is absolutely fine!

**Reply by jrouwe** [reply]
This has been released now.

**Comment by LeXXik** [comment]
Tried to tackle the debug renderer, using the `ShapeGetTriangles`. It kind of works :) Couple of issues:

1. I am iterating the vertices and simply drawing a line from one vertex to another. It kind of works, but not pretty, as I am ignoring indices. Was curious if there is a way to get the indices, so the lines could be drawn in correct order. Not a big deal if not.

![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/01aed17f-ddb1-4583-b4c9-9e030bdd41f6)

2. Kind of a bigger issue is that I can't use this method to draw more than one shape. Most probably I am doing something wrong, but I always get the same bytes offset for a shape, no matter which shape I am requesting it for:

```js
// if I have 2 shapes, I do for each
const center = shape.GetCenterOfMass();
const triContext = new Jolt.ShapeGetTriangles(shape, aabb, center, quat, scale);
const byteOffset = triContext.GetVerticesData(); 
// in this case I get byteOffset = 21823472 for both shapes
// I noticed the shape pointers are different for both shapes, but center seems to be the same
```

So, when I am trying to get vertices for the next shape, I always start from the same heap offset.

**Reply by LeXXik** [reply]
Thinking about the first one - I think its just me not drawing the third line of a triangle, to close the loop. You can ignore it.

**Reply by jrouwe** [reply]
Can you maybe provide a complete example (e.g. by modifying one of the examples) for point 2? When you new Jolt.ShapeGetTriangles it has its own buffer, so I don't see how that could not work when newing multiple ShapeGetTriangles for different shapes.

**Reply by LeXXik** [reply]
Alright, I think I found the cause. When done reading triangle data for a shape, I destroy the tricontext. I think that is causing a reset? I made a small example - if you comment out the destruction of tricontext in the `getByteOffset` function, then the offset is different for each body:

https://jsfiddle.net/LeXXik/sar36vqf/17/

**Comment by jrouwe** [comment, marked as answer]
```
function getByteOffset(body) {
  const shape = body.GetShape();
  const triContext = new Jolt.ShapeGetTriangles(shape, aabb, shape.GetCenterOfMass(), quat, scale);
  const offset = triContext.GetVerticesData();
  Jolt.destroy(triContext);
  return offset;
}
```

This construct cannot work. The vertex data is owned by the ShapeGetTriangles object, so if you destroy it, its memory is freed so your ```offset``` variable is pointing to freed memory.

**Reply by LeXXik** [reply]
Hmm, so to iterate properly, I am now going through all bodies first, caching the contexts. Once done, I destroy all. This seems to work:
![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/6734dbd7-964e-46d9-b05b-5a120d8807a3)

**Reply by LeXXik** [reply]
For future seekers - don't destroy them, actually. The offset may become incorrect, introducing visual artifacts. Create once and keep until debug draw is disabled.

**Comment by LeXXik** [comment]
Final result with/without depth draw, per object draw, own color per motion type:

https://github.com/jrouwe/JoltPhysics.js/assets/5677782/8a476de1-671f-4675-9091-942b0362d8c5

**Comment by LeXXik** [comment]
I am curious if there is a way to map the currently drawn shape to the body. The issue here, is that the vertices stored in the heap are in local space of the mesh/shape. In order for them to match the body transforms, I do the transformation based on the visual mesh, e.g. using the position/rotation of the visual mesh, I transform the local verts to world space. This generally works alright, but in some cases it can mislead. For example, a dynamic body has been moved to one position in the physics world, while a visual mesh was moved by user script to another. This creates a divergence in transforms. Visually, the debug draw would still match the visual mesh, because it uses its transform as a base, but in reality the physical body is somewhere else. To fix it, I'd need to know world transforms of the physical body instead, but not sure if there is a way to get it from that loop. I suppose I'd leave it as is, and mention it as a limitation, but if you have an idea, I'm interested to hear any.

**Reply by jrouwe** [reply]
>  but not sure if there is a way to get it from that loop

I'm not sure what you mean with 'that loop', but don't you have access to ```Body::GetCenterOfMassTransform``` to get the actual transform of the body in 'that loop'?

**Reply by LeXXik** [reply]
Ah, sorry, it was my own loop. All the needed info is available. Please, ignore my question :D I should stop coding late at night....
